plugin/pkg/scheduler/algorithm/predicates/predicates_test.go

```diff
@@ -1588,6 +1592,256 @@ func TestServiceAffinity(t *testing.T) {
+		ObjectMeta: metav1.ObjectMeta{
+			Name: "hostname",
+			Labels: map[string]string{
+				kubeletapis.LabelInstanceType: "small-node",
+			},
+		},
+	}
+
+	tests := []struct {
+		newPod       *v1.Pod
+		existingPods []*v1.Pod
+		node         *v1.Node
+		fits         bool
+		test         string
+	}{
+		{
+			newPod:       twoVolPod,
+			existingPods: append(onePodList_15, onePodList_15...),
+			node:         standard_1_node,
+			fits:         true,
+			test:         "fits when node capacity >= new pod's GCE volumes",
+		},
+		{
+			newPod:       oneVolPod,
+			existingPods: twoPodList_64,
+			node:         standard_2_node,
+			fits:         false,
+			test:         "doesn't fit when node capacity < new pod's GCE volumes",
+		},
+		{
+			newPod:       twoVolPod,
+			existingPods: append(onePodList_15, twoPodList_64...),
+			node:         highcpu_8_node,
+			fits:         true,
+			test:         "fits when node capacity >= new pod's GCE volumes",
+		},
+		{
+			newPod:       oneVolPod,
+			existingPods: onePodList_15,
+			node:         small_node,
+			fits:         true,
+			test:         "fits when node capacity >= new pod's GCE volumes",
+		},
+	}
+
+	filter := VolumeFilter{
+		FilterVolume: func(vol *v1.Volume) (string, bool) {
+			if vol.GCEPersistentDisk != nil {
+				return vol.GCEPersistentDisk.PDName, true
+			}
+			return "", false
+		},
+		FilterPersistentVolume: func(pv *v1.PersistentVolume) (string, bool) {
+			if pv.Spec.GCEPersistentDisk != nil {
+				return pv.Spec.GCEPersistentDisk.PDName, true
+			}
+			return "", false
+		},
+	}
+	expectedFailureReasons := []algorithm.PredicateFailureReason{ErrMaxVolumeCountExceeded}
+
+	pvInfo := FakePersistentVolumeInfo{}
+
+	pvcInfo := FakePersistentVolumeClaimInfo{}
+
+	for _, test := range tests {
+		pred := NewMaxPDVolumeCountPredicate(filter, gce.MaxPDCount, pvInfo, pvcInfo)
+		nodeInfo := schedulercache.NewNodeInfo(test.existingPods...)
+		nodeInfo.SetNode(test.node)
+		fits, reasons, err := pred(test.newPod, PredicateMetadata(test.newPod, nil), nodeInfo)
+		if err != nil {
+			t.Errorf("%s: unexpected error: %v", test.test, err)
+		}
+		if !fits && !reflect.DeepEqual(reasons, expectedFailureReasons) {
+			t.Errorf("%s: unexpected failure reasons: %v, want: %v", test.test, reasons, expectedFailureReasons)
+		}
+		if fits != test.fits {
+			t.Errorf("%s: expected %v, got %v", test.test, test.fits, fits)
+		}
+	}
+
+	// Test get MaxPDCount value from environment varible
+	envTests := []struct {
+		newPod       *v1.Pod
+		existingPods []*v1.Pod
+		node         *v1.Node
+		envMaxPDs    string
+		fits         bool
+		test         string
+	}{
+		{
+			newPod:       twoVolPod,
+			existingPods: append(onePodList_15, onePodList_15...),
+			node:         standard_1_node,
+			envMaxPDs:    "invalid",
+			fits:         true,
+			test:         "fits when node capacity >= new pod's GCE volumes",
+		},
+		{
+			newPod:       twoVolPod,
+			existingPods: append(onePodList_15, onePodList_15...),
```

Comment:
You should only need 15 volumes here right?

Reply:
I was thinking to test the case that if volumes from different pods are the same, they should not be counted twice. But this is not the right way for that scenario. I fixed it by creating a new list of pod which contains some same volumes with another list of pods.